File: bench/shapes.py
import json
import sys
import os
import random
import time

thisdir=os.path.abspath(os.path.dirname(sys.argv[0]))
cfg=json.load(open(thisdir + '/config.json', 'r'))
n=cfg["n"]

class Shape:
    def __init__(self,name):
        self.name=name
    def draw(self):
        print "drawing: " + self.name

class Circle(Shape):
    def __init__(self,name,radius):
        Shape.__init__(self,name)
        self.radius = radius
    def draw(self):
        print "drawing circle: " + self.name + " r=" + str(self.radius) 

class Square(Shape):
    def __init__(self,name,side):
        Shape.__init__(self,name)
        self.side = side
    def draw(self):
        print "drawing square: " + self.name + " s=" + str(self.side) 

def rand(n):
    return random.randint(0,n-1)

def mkShapes(n):
    shapes = list([])
    for i in xrange(n):
        name = str(rand(10000)) + "(" + str(i) + ")";
        if (i%10) <= 5:
            shapes.append(Circle("circle # " + name,rand(100)))
        else:
            shapes.append(Square("shape # " +name,rand(100)))
    return shapes

def drawShapes(shapes):
    for shape in shapes:
        shape.draw()

start = time.clock()
shapes=mkShapes(n)
end = time.clock()
print "construct: " + str(end - start) + "s"

start = time.clock()
sortedShapes = sorted(shapes,key = lambda shape: shape.name)
end = time.clock()
print "sort: " + str(end - start) + "s"

File: bench/cpp11/src/shapes.cpp
#include <vector>
#include <algorithm>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <stdio.h>

// https://github.com/nlohmann/json/releases
#include "json.hpp"

int rand(int n) { return rand() % n; }

class Shape {
public: typedef std::shared_ptr < Shape > Ptr;
public: std::string name;
public: Shape(const std::string &_name) : name(_name) { }
public: virtual void draw() { std::cout << "drawing: " << name << std::endl; }
};

class Circle: public Shape {
public: double radius;
public: Circle(const std::string &_name, int _radius)
  : Shape(_name), radius(_radius) {}

public: virtual void draw() {
  std::cout << "drawing circle: " << name << " r=" << radius << std::endl;
}
};

class Square: public Shape {
public: double side;
public: Square(const std::string &_name, int _side)
  : Shape(_name), side(_side) {}
public: void draw() {
  std::cout << "drawing square: " << name << " s=" << side << std::endl;  
}
};

class Shapes: public Shape {
public: Shapes() : Shape("shapes") {}
public: std::vector < Ptr > parts;
public: void make(int n) {
  char cbuf[1024];
  parts.resize(n, Ptr(0));
  for (int i=0; i<n; ++i) {
    if ((i%10) <= 5) {
      snprintf(cbuf,sizeof(cbuf),"circle # %d (%d)",rand(10000),i);
      parts[i]=Ptr(new Circle(cbuf,rand(100)));
    } else {
      snprintf(cbuf,sizeof(cbuf),"square # %d (%d)",rand(10000),i);      
      parts[i]=Ptr(new Square("shape # " +name,rand(100)));
    }
  }
}

public: void sort() {
  std::sort(parts.begin(), parts.end(),
            [](const Ptr & a, const Ptr & b)
            { return a->name > b->name; });
}

public: void draw() {
  for (auto shape : parts) {
    shape->draw();
  }
}
  
public: void run(int n) {
  Shapes shapes;
  long t0,t1;

  t0=clock();
  shapes.make(n);
  t1=clock();
  std::cout << "construct: " << double(t1-t0)/CLOCKS_PER_SEC << "s" << std::endl;
  t0=clock();
  shapes.sort();
  t1=clock();
  std::cout << "sort: " << double(t1-t0)/CLOCKS_PER_SEC << "s" << std::endl;    
}


};

int main() {
  nlohmann::json config;
  std::ifstream("../config.json") >> config;
  int n = config["n"];
  Shapes shapes;
  shapes.run(n);
  return 0;
}
